broker/src/main/java/io/camunda/zeebe/broker/system/EmbeddedGatewayService.java

```diff
@@ -7,48 +7,71 @@
-import io.atomix.cluster.messaging.ClusterEventService;
-import io.atomix.cluster.messaging.MessagingService;
+import io.camunda.zeebe.broker.clustering.ClusterServices;
 import io.camunda.zeebe.broker.system.configuration.BrokerCfg;
 import io.camunda.zeebe.gateway.Gateway;
+import io.camunda.zeebe.gateway.Loggers;
 import io.camunda.zeebe.gateway.impl.broker.BrokerClientImpl;
+import io.camunda.zeebe.gateway.impl.stream.JobStreamClient;
+import io.camunda.zeebe.gateway.impl.stream.JobStreamClientImpl;
 import io.camunda.zeebe.scheduler.ActorSchedulingService;
+import io.camunda.zeebe.scheduler.ConcurrencyControl;
 import io.camunda.zeebe.scheduler.future.ActorFuture;
+import io.camunda.zeebe.scheduler.future.ActorFutureCollector;
+import org.agrona.CloseHelper;
 
 public final class EmbeddedGatewayService implements AutoCloseable {
   private final Gateway gateway;
   private final BrokerClientImpl brokerClient;
+  private final JobStreamClient jobStreamClient;
+  private final ConcurrencyControl concurrencyControl;
 
   public EmbeddedGatewayService(
       final BrokerCfg configuration,
       final ActorSchedulingService actorScheduler,
-      final MessagingService messagingService,
-      final ClusterMembershipService membershipService,
-      final ClusterEventService eventService) {
+      final ClusterServices clusterServices,
+      final ConcurrencyControl concurrencyControl) {
+    this.concurrencyControl = concurrencyControl;
     brokerClient =
         new BrokerClientImpl(
             configuration.getGateway().getCluster().getRequestTimeout(),
-            messagingService,
-            membershipService,
-            eventService,
+            clusterServices.getMessagingService(),
+            clusterServices.getMembershipService(),
+            clusterServices.getEventService(),
             actorScheduler);
-    gateway = new Gateway(configuration.getGateway(), brokerClient, actorScheduler);
+    jobStreamClient =
+        new JobStreamClientImpl(actorScheduler, clusterServices.getCommunicationService());
+    gateway =
+        new Gateway(
+            configuration.getGateway(), brokerClient, actorScheduler, jobStreamClient.streamer());
   }
 
   @Override
   public void close() {
-    if (gateway != null) {
-      gateway.stop();
-    }
+    CloseHelper.closeAll(
+        error ->
+            Loggers.GATEWAY_LOGGER.warn(
+                "Error occurred while shutting down embedded gateway", error),
+        gateway,
+        brokerClient,
+        jobStreamClient);
   }
 
   public Gateway get() {
     return gateway;
   }
 
   public ActorFuture<Gateway> start() {
-    brokerClient.start();
+    jobStreamClient.start();
+
+    // before we can add the job stream client as a topology listener, we need to wait for the
+    // topology to be set up, otherwise the callback may be lost
+    final var brokerClientStart =
+        brokerClient.start().stream().collect(new ActorFutureCollector<>(concurrencyControl));
```

Comment:
:question: Should we also add the future returned by `jobStreamClient.start()` to wait until jobStreamClient is started before adding it as topology listener?

Reply:
Ah, yes, I added it initially when the start of the job stream client was synchronous :see_no_evil: